recipes/IWSLT22_lowresource/AST/transformer/train_speecht5.py

```diff
@@ -0,0 +1,365 @@
+        """Gets called when a stage (either training, validation, test) starts."""
+
+        if stage != sb.Stage.TRAIN:
+            self.acc_metric = self.hparams.acc_computer()
+            self.bleu_metric = self.hparams.bleu_computer()
+
+    def on_stage_end(self, stage, stage_loss, epoch):
+        """Gets called at the end of a epoch."""
+        # Compute/store important stats
+        if stage == sb.Stage.TRAIN:
+            self.train_stats = stage_loss
+
+        else:  # valid or test
+            stage_stats = {"loss": stage_loss}
+            stage_stats["ACC"] = self.acc_metric.summarize()
+            stage_stats["BLEU"] = self.bleu_metric.summarize(field="BLEU")
+            stage_stats["BLEU_extensive"] = self.bleu_metric.summarize()
+
+        # log stats and save checkpoint at end-of-epoch
+        if stage == sb.Stage.VALID:
+            current_epoch = self.hparams.epoch_counter.current
+            old_lr_adam, new_lr_adam = self.hparams.lr_annealing_adam(
+                stage_stats["BLEU"]
+            )
+            sb.nnet.schedulers.update_learning_rate(
+                self.adam_optimizer, new_lr_adam
+            )
+
+            self.hparams.train_logger.log_stats(
+                stats_meta={"epoch": current_epoch, "lr_adam": old_lr_adam},
+                train_stats={"loss": self.train_stats},
+                valid_stats=stage_stats,
+            )
+
+            # create checkpoing
+            meta = {"BLEU": stage_stats["BLEU"], "epoch": current_epoch}
+            name = "checkpoint_epoch" + str(current_epoch)
+
+            self.checkpointer.save_and_keep_only(
+                meta=meta, name=name, num_to_keep=5, max_keys=["BLEU"]
+            )
+
+        elif stage == sb.Stage.TEST:
+            self.hparams.train_logger.log_stats(
+                stats_meta={"Epoch loaded": self.hparams.epoch_counter.current},
+                test_stats=stage_stats,
+            )
+
+
+# Define custom data procedure
+def dataio_prepare(hparams, tokenizer):
+    """This function prepares the datasets to be used in the brain class.
+    It also defines the data processing pipeline through user-defined functions.
+    """
+
+    # Define audio pipeline. In this case, we simply read the path contained
+    # in the variable wav with the audio reader.
+    @sb.utils.data_pipeline.takes("path")
+    @sb.utils.data_pipeline.provides("sig")
+    def audio_pipeline(wav):
+        """Load the audio signal. This is done on the CPU in the `collate_fn`."""
+        sig = sb.dataio.dataio.read_audio(wav)
+        return sig
+
+    @sb.utils.data_pipeline.takes("path")
+    @sb.utils.data_pipeline.provides("sig")
+    def sp_audio_pipeline(wav):
+        """Load the audio signal. This is done on the CPU in the `collate_fn`."""
+        sig = sb.dataio.dataio.read_audio(wav)
+        sig = sig.unsqueeze(0)
+        sig = hparams["speed_perturb"](sig)
+        sig = sig.squeeze(0)
+        return sig
```

Comment:
hmmm could you please explain why are you doing speed_perturb outside of the compute_forward? We are usually doing it there